refactor(docker): Enhance Zebra configuration options and entrypoint logic

gustavovalverde opened the pull request:
## Motivation

We had some issues with how we were dealing with configuration files:
- If the user mounted a configuration file, zebra would (try to) overwrite it. This could cause odd behaviors like:
  - #9325 
- When mounting a configuration file using the [docker compose configs](https://docs.docker.com/reference/compose-file/configs/) this would cause an error as the mounted directory is _read-only_
- We should be able to differentiate between a _user-provided_ configuration file, our _default_ configuration file (mainly used with docker compose), and if there's no configuration file provided, so we can build it _on-the-fly_ with the provided variables by the user.

A possible approach to fix this was using a temporal directory, and using the resulting config from that directory instead from the original mounted dir, but this was also problematic if the temporal directory was mounted as read-only from the host, which also led to the decision of not using `sed` and trying to replace the content of the file or creating a new file in a location that could potentially be `read-only` by external host limitations.

Fixes #9325 

## Solution

**Key changes:**
- Do not mount a default configuration file in our Dockerfile (`default-zebra-config.toml`), to allow more flexibility: default configuration build, default config file mounting, user-provided config file mounting with a custom `$ZEBRA_CONF_PATH`
  - Validate if the user is mounting its own file (this requires them to set `$ZEBRA_CONF_PATH` to a valid mounted file)
  - If the user doesn't set `ZEBRA_CONF_PATH` then validate if the default `${HOME}/.config/zebrad.toml` is mounted (commonly provided with our docker compose https://github.com/ZcashFoundation/zebra/blob/ac25192afcb1c3280ff32d2bb2e75f91f1d9ceb0/docker/docker-compose.yml#L23)
  - Otherwise, create a zebra configuration file based on the variables set by the user (or using the default values)
- Do not try to overwrite mounted files, dynamic configuration should only happen `on-the-fly` and must be avoided if a configuration file was provided (this changes `sed` to `cat` as using `sed` is no longer required)

**Other (required) changes**
- `exec_as_user` must validate if the actual user is root, otherwise running `gosu` will fail when that's not the case
- If cache or other directories were configured with the environment variables, make sure they exists and change their ownership to the user running `zebrad`.
- Do not mount the `.cookies` directory in the same directory as the `.cache` as mounting a cached state would override the cookies directory.

### Tests

All tests should be passing, and building with the docker compose, with our without a mount file, should give the expected configuration results.

### Follow-up Work

Validate which other configurations should be added using variables

### PR Checklist

<!-- Check as many boxes as possible. -->

- [x] The PR name is suitable for the release notes.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.
- [ ] If the PR shouldn't be in the release notes, it has the
      `C-exclude-from-changelog` label.

gustavovalverde committed:
refactor(docker): Enhance Zebra configuration options and entrypoint logic

- Introduced multiple methods for configuring Zebra, including custom config file paths and automatic generation from environment variables.
- Updated entrypoint script to handle configuration more robustly, checking for existing files and generating defaults as needed.
- Improved documentation in `docker.md` to clarify configuration options and their precedence.
- Adjusted environment variable handling in `docker/.env` for better clarity and functionality.
- Refactored `prepare_conf_file` to create a complete configuration based on environment variables, streamlining the setup process.

upbqdn commented on docker/default-zebra-config.toml:
```diff
@@ -23,7 +23,7 @@ cache_dir = "/home/zebra/.cache/zebra"
 # listen_addr = "0.0.0.0:18232" # Testnet
 
-cookie_dir = "/home/zebra/.cache/zebra"
+cookie_dir = "/home/zebra/.config/cookie"
```
This is a breaking change. We need to update the existing docs and add new since the location differs from where Zebra stores the cookie by default.

upbqdn replied:
We reverted this change.

upbqdn commented on book/src/user/docker.md:
```diff
@@ -59,19 +59,22 @@ All available Cargo features are listed at
+The entrypoint script follows this sequence:
+
+1. If `ZEBRA_CONF_PATH` is set and a file exists at that path, it uses that file
+2. If `ZEBRA_CONF_PATH` is not set but a default config exists at `${HOME}/.config/zebrad.toml`, it uses that file
```
It won't be the default config if users mount their own.

```suggestion
2. If `ZEBRA_CONF_PATH` is not set but a file exists at `${HOME}/.config/zebrad.toml`, it uses that file
```

gustavovalverde replied:
The user should mount their file and set `ZEBRA_CONF_PATH` if they want to use their own custom configuration, otherwise the default config (`default-zebra-config.toml`) will be used if it's mounted. That's now the default approach in the `docker-compose.yml`.

Certainly, the user can set `ZEBRA_CONF_PATH` and mount their file to the default location`${HOME}/.config/zebrad.toml`, but that case would fall into the _user-provided_ condition.

Recent change to make this clearer: [`be37ecb` (#9344)](https://github.com/ZcashFoundation/zebra/pull/9344/commits/be37ecb853637ec7494d9202060a50d112383ba3)

upbqdn commented on docker/entrypoint.sh:
```diff
@@ -6,113 +6,125 @@
 #
-# - `$ZEBRA_CONF_PATH` must point to a Zebra conf file.
+# - `$ZEBRA_CONF_PATH` can point to an existing Zebra config file, or if not set,
+#   the script will look for a default config at ${HOME}/.config/zebrad.toml,
```
```suggestion
#   the script will look for a config at ${HOME}/.config/zebrad.toml,
```

gustavovalverde replied:
https://github.com/ZcashFoundation/zebra/pull/9344#discussion_r2033143765

upbqdn commented on docker/entrypoint.sh:
```diff
@@ -255,12 +267,39 @@ run_tests() {
+else
+  if [[ -f "${HOME}/.config/zebrad.toml" ]]; then
+    echo "ZEBRA_CONF_PATH was not set."
+    echo "Using default config at ${HOME}/.config/zebrad.toml"
```
```suggestion
    echo "Using the config at ${HOME}/.config/zebrad.toml"
```

gustavovalverde replied:
https://github.com/ZcashFoundation/zebra/pull/9344#discussion_r2033143765

upbqdn commented on docker/entrypoint.sh:
```diff
@@ -255,12 +267,39 @@ run_tests() {
+    echo "Using default config at ${HOME}/.config/zebrad.toml"
+    ZEBRA_CONF_PATH="${HOME}/.config/zebrad.toml"
+  else
+    echo "ZEBRA_CONF_PATH was not set and no default config found at ${HOME}/.config/zebrad.toml"
```
```suggestion
    echo "ZEBRA_CONF_PATH was not set and no config found at ${HOME}/.config/zebrad.toml"
```

gustavovalverde replied:
https://github.com/ZcashFoundation/zebra/pull/9344#discussion_r2033143765

upbqdn commented on docker/.env:
```diff
@@ -1,30 +1,60 @@
 # FEATURES=""
 
+# Sets the listen address and port for Prometheus metrics.
+# Only active when the 'prometheus' feature is enabled.
```
We'll make the `prometheus` feature enabled by default soon.

```suggestion
```

upbqdn commented on docker/.env:
```diff
@@ -1,30 +1,60 @@
+# Sets the listen address for Zebra's peer-to-peer network connections.
+#
+# ZEBRA_LISTEN_ADDR="0.0.0.0"
+
```
I don't think users should be presented with this option since using any other address will make the endpoint inaccessible outside Docker.

gustavovalverde replied:
Fixed with [`9e336f6` (#9344)](https://github.com/ZcashFoundation/zebra/pull/9344/commits/9e336f65d5647c8d0fb6901d6df1e5fc697275c2)

upbqdn commented on docker/.env:
```diff
@@ -1,30 +1,60 @@
+# Sets a custom directory for the lightwalletd cache.
+#
+# LWD_CACHE_DIR="/home/zebra/.cache/lwd"
+
```
Is there any use-case for `LWD_CACHE_DIR`?

```suggestion
```

gustavovalverde replied:
Fixed with [`9e336f6` (#9344)](https://github.com/ZcashFoundation/zebra/pull/9344/commits/9e336f65d5647c8d0fb6901d6df1e5fc697275c2)

upbqdn commented on docker/.env:
```diff
@@ -1,30 +1,60 @@
+# Sets the path to a custom Zebra config file. If not set, Zebra will look for a config at
+# ${HOME}/.config/zebrad.toml or generate one using environment variables below.
+# ! Setting ZEBRA_CONF_PATH will make most of the following environment variables ineffective.
+# TODO: zebrad should be able to read from both the config file and environment variables.
```
Since this is a user-facing config file, I expect the TODO will be confusing.

```suggestion
```

upbqdn reviewed (commented):
@gustavovalverde, have you tested the solution? It looks like some parts don't work. For example, setting the RPC port.

If I set `ZEBRA_RPC_PORT=28232`, the script configures Zebra to use

```toml
[rpc]
listen_addr = "0.0.0.0:8232"
```

which is not the desired port.

upbqdn commented:
The issue is similar to what we had before the refactor:

Even setting `ZEBRA_RPC_PORT=catchthisbug` makes Zebra run just fine. It will be hard to debug this if we have something wrong, but things will run anyway.

upbqdn commented:
I wonder if there's a way we could test these things. It should be quick to write a bunch of basic tests if we had a suitable framework for them.

gustavovalverde commented:
> I wonder if there's a way we could test these things. It should be quick to write a bunch of basic tests if we had a suitable framework for them.

I'll start a PR tomorrow just for that. There's just way too many edge-cases that I can forget and can go wrong with this changes, I'll just start building a test suite and make this PR dependent on that.

I'll start with https://github.com/ZcashFoundation/zebra/issues/9037

I'll put this in draft, and let's revisit once the test suite is good enough

gustavovalverde committed:
fix(entrypoint): use correct `ZEBRA_RPC_PORT` variable

This also removes the logic to set the default port for RPC as this was introducing a buggy logic

gustavovalverde committed:
fix(docker): remove non-essential variables and set correct defaults

gustavovalverde committed:
chore(docker): use the default config for docker-compose as the default approach

gustavovalverde committed:
Apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

gustavovalverde commented:
I created https://github.com/ZcashFoundation/zebra/pull/9388 to be able to test different scenarios from these changes, and made the fixes here to correctly support them.

This should be ready now @upbqdn

upbqdn commented on docker/docker-compose.yml:
```diff
@@ -15,13 +15,12 @@ services:
-    #     target: /home/zebra/.config/zebrad.toml
-    #
+    # ! Comment out the `configs` mapping below to not use our default configuration file.
+    # ! This will use the environment variables in the `.env` file.
```
From the user's perspective, it's not clear to me what _"This"_ refers to. Is it the comment above or the configs below?

gustavovalverde replied:
Fixed here: [chore: better comment for user understanding](https://github.com/ZcashFoundation/zebra/pull/9344/commits/63f4c6795568f676c788fb785283a6f11ad11af8)

gustavovalverde commented:
@upbqdn I actually didn't realize I left #9388 targeting this branch. I did it just to be able to rebase this one and test the #9388. I'm planning to revert the merge, and reopen the PR, so we can merge them individually.

This is because we're going to squash the commits, and then the title and PR content will be confusing, and unrelated to most of the changes.

gustavovalverde committed:
chore: better comment for user understanding

gustavovalverde commented:
@upbqdn This PR is clean now, and ready for review/approval

upbqdn commented on docker/entrypoint.sh:
```diff
@@ -255,12 +258,39 @@ run_tests() {
+    echo "Preparing a default one..."
+    ZEBRA_CONF_PATH="${HOME}/.config/zebrad.toml"
+    mkdir -p "$(dirname "${ZEBRA_CONF_PATH}")"
+    prepare_conf_file
```
###### non-blocking

This will create the config file and its parent dir owned by root:

```bash
root@1cbf7683b54d:~/.config# pwd
/home/zebra/.config
root@1cbf7683b54d:~/.config# ls -lah
total 4.0K
drwxr-xr-x 1 root  root   34 Apr 14 14:10 .
drwx------ 1 zebra zebra  26 Apr 14 14:10 ..
drwxr-xr-x 1 zebra zebra   0 Apr 14 14:10 cookie
-rw-r--r-- 1 root  root  149 Apr 14 14:10 zebrad.toml
```

gustavovalverde replied:
Fixed here: [fix(entrypoint): set permissions correctly for parent directories](https://github.com/ZcashFoundation/zebra/pull/9344/commits/e45c951b15993c83b1c73654779c83aa1292fa31)

upbqdn replied:
This solution fixes the problem only for immediate parent dirs instead of the full path.

This is not a blocking bug, but it might bite us in the future.

upbqdn replied:
Note that the config file is also owned by root.

gustavovalverde replied:
I changed the specific use case when a file is generated by the entrypoint and not mounted, as trying to own a mounted file has caused issues previously: [fix(entrypoint): chown the zebra config file when](https://github.com/ZcashFoundation/zebra/pull/9344/commits/94c5a3e6a5c97fe3bcc0f684bef3a821f7cdfbf7)

upbqdn commented on docker/entrypoint.sh:
```diff
@@ -6,113 +6,116 @@
 
-  echo "${ZEBRA_CONF_PATH}"
+  mkdir -p "${dir}" || exit_error "Failed to create directory: ${dir}"
+  chown -R "${UID}:${GID}" "${dir}" || exit_error "Failed to secure directory: ${dir}"
```
###### non-blocking

This will create parent dirs such as `.cache` owned by root:

```bash
root@1cbf7683b54d:~# pwd
/home/zebra
root@1cbf7683b54d:~# ls -lah
total 12K
drwx------ 1 zebra zebra   26 Apr 14 14:10 .
drwxr-xr-x 1 root  root    10 Apr  9 13:03 ..
-rw-r--r-- 1 zebra zebra  220 Apr  9 13:03 .bash_logout
-rw-r--r-- 1 zebra zebra 3.5K Apr  9 13:03 .bashrc
drwxr-xr-x 1 root  root    16 Apr 14 14:10 .cache
drwxr-xr-x 1 root  root    34 Apr 14 14:10 .config
-rw-r--r-- 1 zebra zebra  807 Apr  9 13:03 .profile
```

gustavovalverde replied:
Fixed here: [fix(entrypoint): set permissions correctly for parent directories](https://github.com/ZcashFoundation/zebra/pull/9344/commits/e45c951b15993c83b1c73654779c83aa1292fa31)

upbqdn commented on book/src/user/docker.md:
```diff
@@ -59,19 +59,22 @@ All available Cargo features are listed at
-provides full config capabilities.
+To configure Zebra, you have three options:
+
+1. Edit the `docker/default-zebra-config.toml` config file and uncomment the `configs` mapping in `docker/docker-compose.yml` so your config takes effect.
```
This documentation is outdated.

gustavovalverde replied:
Fixed here: [chore: align documentation and implementation](https://github.com/ZcashFoundation/zebra/pull/9344/commits/2520086bd2effaefe7ee2ddf7caa684d3d897904)

upbqdn commented on book/src/user/docker.md:
```diff
@@ -59,19 +59,22 @@ All available Cargo features are listed at
+
+1. Edit the `docker/default-zebra-config.toml` config file and uncomment the `configs` mapping in `docker/docker-compose.yml` so your config takes effect.
+2. Provide a path to your own config file via the `ZEBRA_CONF_PATH` environment variable.
+3. Let Zebra generate a default config file automatically from environment variables (if neither of the above methods is used).
```
Zebra doesn't generate the config. We'd use `zebrad generate` for that (and it wouldn't consider env vars).

gustavovalverde replied:
Fixed here: [chore: align documentation and implementation](https://github.com/ZcashFoundation/zebra/pull/9344/commits/2520086bd2effaefe7ee2ddf7caa684d3d897904)

gustavovalverde committed:
fix(entrypoint): set permissions correctly for parent directories

gustavovalverde committed:
chore: align documentation and implementation

upbqdn commented on book/src/user/docker.md:
```diff
@@ -59,49 +59,47 @@ All available Cargo features are listed at
+To configure Zebra using Docker, you have a few options, processed in this order:
+
+1. **Provide a specific config file path:** Set the `ZEBRA_CONF_PATH` environment variable to point to your config file within the container.
+2. **Use the default config file:** If `ZEBRA_CONF_PATH` is not set, Zebra will look for a config file at `/home/zebra/.config/zebrad.toml`. You can mount your custom config file to this location, for example, by uncommenting the `configs` mapping in `docker/docker-compose.yml`.
```
The reason I mentioned the documentation was outdated was that this PR uncomments the `configs` mapping in `docker/docker-compose.yml`, but the documentation was still saying that users should uncomment the mapping. This update says it again.

gustavovalverde replied:
Fixed here: [chore: clearer and aligned comments with docker-compose](https://github.com/ZcashFoundation/zebra/pull/9344/commits/c5610648a150edb74a5b10309df978d69ca23dfb)

gustavovalverde committed:
chore: clearer and aligned comments with docker-compose

upbqdn commented:
We should update the cookie path here as well: https://github.com/zcashfoundation/zebra/blob/c5610648a150edb74a5b10309df978d69ca23dfb/docker/default-zebra-config.toml#L26.

gustavovalverde commented:
> We should update the cookie path here as well: [zcashfoundation/zebra@`c561064`/docker/default-zebra-config.toml#L26](https://github.com/zcashfoundation/zebra/blob/c5610648a150edb74a5b10309df978d69ca23dfb/docker/default-zebra-config.toml#L26).

This seems correct: `cookie_dir = "/home/zebra/.cache/zebra"`

All other locations and docs are pointing to this same dir; except for the commented out $ZEBRA_COOKIE_DIR here https://github.com/ZcashFoundation/zebra/pull/9344/files#diff-8f78f3866f02305f1df51579c939174b2c53e8295ad13bd9d2ab763176a792c1R24 which refers to a custom location, as the comment above also implies. 

There are comments referring to the cookie file itself, and those are pointing to `/home/zebra/.cache/zebra/.cookie`

gustavovalverde committed:
fix(entrypoint): chown the zebra config file when `ZEBRA_CONF_PATH` is not set

upbqdn commented:
utACK

One simplification I thought of was if we could omit the first config option, and only go with the static location at `/home/zebra/.config/zebrad.toml` we already use in the container. The script would then check if there's a file at that location, and if not, it would generate one. Omitting custom config locations would also simplify the docs quite a bit.

gustavovalverde commented:
> Omitting custom config locations would also simplify the docs quite a bit.

I agree with this, but there are edge-cases we have to consider and we might break those https://github.com/ZcashFoundation/zebra/issues/3432

Unfortunately, this depend on the host and the security policies set on it for docker.
